apple-codesign/src/windows.rs

```diff
@@ -0,0 +1,736 @@
+///
+/// `thumbprint` specifies the SHA1 thumbprint of the certificate to search for.
+/// You can find this in `certmgr.msc` or `certlm.msc` by clicking on the certificate in
+/// question and looking for `Thumbprint` under the `Details` tab.
+pub fn windows_store_find_certificate_chain(
+    store_name: StoreName,
+    thumbprint: &str,
+) -> Result<Vec<CapturedX509Certificate>, AppleCodesignError> {
+    // We look for the code signing certificate in the MY store.
+    let user_certs = find_certificates(store_name, StoreType::MY)?;
+
+    // Now search for the requested start certificate and pull the thread until
+    // we get to a self-signed certificate.
+    let start_cert: &CapturedX509Certificate = user_certs
+        .iter()
+        .find_map(|cert| {
+            if let Ok(digest) = cert
+                .captured
+                .sha1_fingerprint()
+            {
+                // Convert the Digest into a byte array
+                let digest_bytes = digest.as_ref();
+
+                // Format each byte as a two-character hex string
+                let digest_hex: String = digest_bytes
+                    .iter()
+                    .map(|byte| format!("{:02x}", byte))
+                    .collect::<String>()
+                    .to_lowercase();
+
+                if digest_hex == thumbprint.to_lowercase() {
+                    Some(&cert.captured)
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        })
+        .ok_or_else(|| AppleCodesignError::CertificateNotFound(format!("Thumbprint={}", thumbprint)))?;
+
+    let mut chain = vec![start_cert.clone()];
+    let mut last_issuer_name = start_cert.issuer_name();
+    
+    if start_cert.issuer_name() == start_cert.subject_name() {
+        // Self signed. Stop the chain.
+        return Ok(chain);
+    }
+
+    // We look for the certificate chain in the CA and ROOT Stores.
+    let intermediate_ca_certs = find_certificates(store_name, StoreType::CA)?;
+    let root_ca_certs = find_certificates(store_name, StoreType::ROOT)?;
+    let ca_certs = intermediate_ca_certs
+        .into_iter()
+        .chain(root_ca_certs.into_iter())
+        .collect::<Vec<_>>();
+    loop {
+        let issuer = ca_certs.iter().find_map(|cert| {
+            if cert.captured.subject_name() == last_issuer_name {
+                Some(&cert.captured)
+            } else {
+                None
+            }
+        });
+
+        if let Some(issuer) = issuer {
+            chain.push(issuer.clone());
+
+            // Self signed. Stop the chain so we don't infinite loop.
+            if issuer.subject_name() == issuer.issuer_name() {
+                break;
+            } else {
+                last_issuer_name = issuer.issuer_name();
+            }
+        } else {
+            // Couldn't find issuer. Stop the search.
+            break;
+        }
+    }
```

Comment:
This logic is buggy in that it assumes that comparing the certificate names is sufficient to finding the signing certificate. If there are multiple certificates with the same name, this could return an inappropriate choice.

Fortunately, `CapturedX509Certificate` has a `resolve_signing_chain()` method where you can pass a set of certificates and it attempts to pull out the chain. Just collect all the certs from the stores and call this method.

Reply:
Totally missed `resolve_signing_chain`, thanks! Worth mentioning that I took this code from `macos.rs`, so maybe this should be fixed there as well.